File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/model/inheritance/RolePermissionCountResolver.java
package org.babyfish.jimmer.sql.model.inheritance;

import org.babyfish.jimmer.lang.Ref;
import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.TransientResolver;
import org.babyfish.jimmer.sql.ast.tuple.Tuple2;

import java.sql.Connection;
import java.util.*;

public class RolePermissionCountResolver implements TransientResolver<Long, Integer> {

    private final JSqlClient sqlClient;

    public RolePermissionCountResolver(JSqlClient sqlClient) {
        this.sqlClient = sqlClient;
        sqlClient.getTriggers().addAssociationListener(RoleProps.PERMISSIONS, e -> {
            sqlClient.getCaches().getPropertyCache(RoleProps.PERMISSION_COUNT).delete(e.getSourceId());
        });
    }

    @Override
    public Map<Long, Integer> resolve(Collection<Long> roleIds, Connection con) {
        PermissionTable permission = PermissionTable.$;
        List<Tuple2<Long, Long>> tuples = sqlClient
                .createQuery(permission)
                .where(permission.role().id().in(roleIds))
                .groupBy(permission.role().id())
                .select(
                            permission.role().id(),
                            permission.count()
                )
                .execute(con);
        return Tuple2.toMap(tuples, Long::intValue);
    }

    @Override
    public Ref<SortedMap<String, Object>> getParameterMapRef() {
        return sqlClient.getFilters().getTargetParameterMapRef(RoleProps.PERMISSIONS);
    }
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/model/inheritance/RolePermissionCountResolver.kt
package org.babyfish.jimmer.sql.kt.model.inheritance

import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.kt.KTransientResolver
import org.babyfish.jimmer.sql.kt.ast.expression.count
import org.babyfish.jimmer.sql.kt.ast.expression.valueIn
import java.sql.Connection
import java.util.*

class RolePermissionCountResolver(
    private val sqlClient: KSqlClient
) : KTransientResolver.Parameterized<Long, Int> {

    init {
        sqlClient.triggers.addAssociationListener(Role::permissions) {
            sqlClient
                .caches
                .getPropertyCache<Long, Int>(Role::permissionCount)
                ?.delete(it.sourceId as Long)
        }
    }

    override fun resolve(ids: Collection<Long>, con: Connection): Map<Long, Int> =
        sqlClient
            .createQuery(Permission::class) {
                where(table.role.id valueIn ids)
                groupBy(table.role.id)
                select(
                    table.role.id,
                    count(table)
                )
            }
            .execute(con)
            .associateBy({
                it._1
            }) {
                it._2.toInt()
            }

    override fun getParameters(): SortedMap<String, Any>? =
        sqlClient
            .filters
            .getParameterizedTargetFilter(Role::permissions)
            ?.getParameters()
}